packages/web3-eth/src/eth_tx.ts

```diff
@@ -0,0 +1,315 @@
+): Transaction<ReturnType> {
+	if (overrideMethod !== undefined) return overrideMethod(transaction);
+	const formattedTransaction = {
+		...transaction,
+		value:
+			transaction.value === '0x' ? '0x' : convertToValidType(transaction.value, desiredType),
+		gas: convertToValidType(transaction.gas, desiredType),
+		gasPrice: convertToValidType(transaction.gasPrice, desiredType),
+		type: convertToValidType(transaction.type, desiredType),
+		maxFeePerGas: convertToValidType(transaction.maxFeePerGas, desiredType),
+		maxPriorityFeePerGas: convertToValidType(transaction.maxPriorityFeePerGas, desiredType),
+		nonce: convertToValidType(transaction.nonce, desiredType),
+		chainId: convertToValidType(transaction.chainId, desiredType),
+		gasLimit: convertToValidType(transaction.gasLimit, desiredType),
+		v: convertToValidType(transaction.v, desiredType),
+		common: {
+			...transaction.common,
+			customChain: {
+				...transaction.common?.customChain,
+				networkId: convertToValidType(
+					transaction.common?.customChain?.networkId,
+					desiredType,
+				),
+				chainId: convertToValidType(transaction.common?.customChain?.chainId, desiredType),
+			},
+		},
+	};
+	// TODO - TSC is complaining that ReturnType could be instantiated with an
+	// arbitrary type which could be unrelated to 'string | number | bigint | undefined'
+	return formattedTransaction as unknown as Transaction<ReturnType>;
+}
+
+export const detectTransactionType = (
+	transaction: Transaction,
+	overrideMethod?: (transaction: Transaction) => Numbers | undefined,
+): Numbers | undefined => {
+	if (overrideMethod !== undefined) return overrideMethod(transaction);
+	if (transaction.type !== undefined) return transaction.type;
+
+	if (transaction.maxFeePerGas !== undefined || transaction.maxPriorityFeePerGas !== undefined)
+		return '0x2';
+	if (transaction.hardfork === 'london') return '0x2';
+	if (transaction.common?.hardfork === 'london') return '0x2';
+
+	if (transaction.accessList !== undefined) return '0x1';
+	if (transaction.hardfork === 'berlin') return '0x1';
+	if (transaction.common?.hardfork === 'berlin') return '0x1';
+
+	return undefined;
+};
+
+const validateCustomChainInfo = (transaction: Transaction) => {
+	if (transaction.common !== undefined) {
+		if (transaction.common.customChain === undefined) throw new MissingCustomChainError();
+		if (transaction.common.customChain.chainId === undefined)
+			throw new MissingCustomChainIdError();
+		if (
+			transaction.chainId !== undefined &&
+			transaction.chainId !== transaction.common.customChain.chainId
+		)
+			throw new ChainIdMismatchError({
+				txChainId: transaction.chainId,
+				customChainId: transaction.common.customChain.chainId,
+			});
+	}
+};
+
+const validateChainInfo = (transaction: Transaction) => {
+	if (
+		transaction.common !== undefined &&
+		transaction.chain !== undefined &&
+		transaction.hardfork !== undefined
+	)
+		throw new CommonOrChainAndHardforkError();
```

Comment:
I think user can define all these three together, or is there any reference I am missing.

Reply:
This is [1.x behavior](https://github.com/ChainSafe/web3.js/blob/5df6818ee4a0a30fb64dd3e94d97964826489b7f/packages/web3-eth-accounts/src/index.js#L316)